finat/tensor_product.py

```diff
@@ -162,6 +164,64 @@ def point_evaluation(self, order, point, entity=None):
 
         return self._merge_evaluations(factor_results)
 
+    @property
+    def dual_basis(self):
+        # Outer product the dual bases of the factors
+        qs, pss = zip(*(f.dual_basis for f in self.factors))
+        # TODO: check if this Q_is_identity calculation is correct
+        self.Q_is_identity = all(f.Q_is_identity for f in self.factors)
+        ps = TensorPointSet(pss)
+        tuples_of_q_shape_indices = tuple(gem.indices(len(q.shape)) for q in qs)
+        q_muls = reduce(operator.mul, (q[i] for q, i in zip(qs, tuples_of_q_shape_indices)))
+        # Flatten to get final shape indices
+        q_shape_indices = tuple(index for tuple_of_indices in tuples_of_q_shape_indices for index in tuple_of_indices)
+        Q = gem.ComponentTensor(q_muls, q_shape_indices)
+        return Q, ps
+
+    def dual_evaluation(self, fn):
+        if not hasattr(fn, 'factors'):
+
+            Q, x = self.dual_basis
+
+            #
+            # EVALUATE fn AT x
+            #
+            expr = fn(x)
+
+            #
+            # TENSOR CONTRACT Q WITH expr
+            #
+
+            # NOTE: any shape indices in the expression are because the
+            # expression is tensor valued.
+            # NOTE: here the first num_factors rows of Q are node sets (1 for
+            # each factor)
+            # TODO: work out if there are any other cases where the basis
+            # indices in the shape of the dual basis tensor Q are more than
+            # just the first shape index (e.g. with EnrichedElement)
+            num_factors = total_num_factors(self.factors)
+            assert expr.shape == Q.shape[num_factors:]
+            expr_shape_indices = tuple(gem.Index(extent=ex) for ex in expr.shape)
+            basis_indices = tuple(gem.Index(extent=ex) for ex in Q.shape[:num_factors])
+            # TODO: Work out how to deal with identity Q in this case.
+            dual_evaluation_indexed_sum = gem.optimise.make_product((Q[basis_indices + expr_shape_indices], expr[expr_shape_indices]), x.indices+expr_shape_indices)
+
+            return dual_evaluation_indexed_sum, basis_indices
```

Comment:
I think the generic implementation is:
```
shape_indices = gem.indices(len(expr.shape))
basis_indices = gem.indices(len(Q.shape) - len(expr.shape))
Qf = IndexSum(Q[basis_indices + expr_shape_indices] * expr[expr_shape_indices], x.indices + shape_indices)
Qf = gem.optimise.contraction(Qf)
```

Reply:
yes it seems you are correct. I'll make all of them follow this

Reply:
(i.e. just use this in `FiniteElementBase`)

Reply:
I tell a lie it doesn't work for tensorfiniteelement in all cases

Reply:
Can you show an example? I have been playing round with this a little.

Reply:
Try `tests/regression/test_interpolate_vs_project.py::test_interpolate_vs_project[square-VectorFunctionSpace(BDM2)]` in firedrake: trying to work out what the issue is

Reply:
It's either to do with getting the delta elimination to work properly by having the right indices or it's to do with not accounting for `self.base_element.get_value_indices()` in the generic case I think

Reply:
I think the issue is that we're contracting over all the expression shape indices which we shouldn't do here - some of them are actually part of the basis

Reply:
I think that's right. We're contracting over enough indices, just some of the wrong ones.

Reply:
Maybe we have to have the dual_evaluation separate for this case. Perhaps this diff, which I think handles the `transpose` argument correctly?
```
diff --git a/finat/tensorfiniteelement.py b/finat/tensorfiniteelement.py
index 839a009..2db8868 100644
--- a/finat/tensorfiniteelement.py
+++ b/finat/tensorfiniteelement.py
@@ -5,8 +5,6 @@ import numpy
 import gem
 
 from finat.finiteelementbase import FiniteElementBase
-from finat.tensor_product import total_num_factors
-from finat.cube import FlattenedDimensions
 
 
 class TensorFiniteElement(FiniteElementBase):
@@ -124,59 +122,53 @@ class TensorFiniteElement(FiniteElementBase):
 
     @property
     def dual_basis(self):
-
+        if self._transpose:
+            raise NotImplementedError("Not implemented for transposed TFE")
         base = self.base_element
         Q, points = base.dual_basis
 
         # Suppose the tensor element has shape (2, 4)
         # These identity matrices may have difference sizes depending the shapes
         # tQ = Q ⊗ 𝟙₂ ⊗ 𝟙₄
-        Q_shape_indices = gem.indices(len(Q.shape))
-        i_s = tuple(gem.Index(extent=d) for d in self._shape)
-        j_s = tuple(gem.Index(extent=d) for d in self._shape)
-        # we need one delta for each shape axis
-        deltas = reduce(gem.Product, (gem.Delta(i, j) for i, j in zip(i_s, j_s)))
-        # TODO Need to check how this plays with the transpose argument to TensorFiniteElement.
-        tQ = gem.ComponentTensor(Q[Q_shape_indices]*deltas, Q_shape_indices + i_s + j_s)
+        scalar_i = self.base_element.get_indices()
+        scalar_vi = self.base_element.get_value_indices()
+        tensor_i = tuple(gem.Index(extent=d) for d in self._shape)
+        tensor_vi = tuple(gem.Index(extent=d) for d in self._shape)
+        # Couple new basis function and value indices
+        deltas = reduce(gem.Product, (gem.Delta(j, k)
+                                      for j, k in zip(tensor_i, tensor_vi)))
+        if self._transpose:
+            index_ordering = tensor_i + scalar_i + tensor_vi + scalar_vi
+        else:
+            index_ordering = scalar_i + tensor_i + tensor_vi + scalar_vi
 
-        return tQ, points
+        Qi = Q[scalar_i + scalar_vi]
+        Q = gem.ComponentTensor(Qi*deltas, index_ordering)
+        return Q, points
 
     def dual_evaluation(self, fn):
-
         Q, x = self.dual_basis
         expr = fn(x)
-
-        # TODO: Add shortcut (if relevant) for Q being identity tensor
-
+        gem.optimise.contraction(expr)
         # NOTE: any shape indices in the expression are because the expression
         # is tensor valued.
-        assert set(expr.shape) == set(self.value_shape)
-
-        base_value_indices = self.base_element.get_value_indices()
-
-        # Reconstruct the indices from the deltas in dual_basis above and
-        # contract Q with expr
-        Q_shape_indices = tuple(gem.Index(extent=ex) for ex in Q.shape)
-        expr_contraction_indices = tuple(gem.Index(extent=d) for d in self._shape)
-        basis_tensor_indices = tuple(gem.Index(extent=d) for d in self._shape)
-        # NOTE: here the first num_factors rows of Q are node sets (1 for
-        # each factor)
-        # TODO: work out if there are any other cases where the basis
-        # indices in the shape of the dual basis tensor Q are more than
-        # just the first shape index (e.g. with EnrichedElement)
-        if hasattr(self.base_element, 'factors'):
-            num_factors = total_num_factors(self.base_element.factors)
-        elif isinstance(self.base_element, FlattenedDimensions):
-            # Factor might be a FlattenedDimensions which introduces
-            # another factor without having a factors property
-            num_factors = 2
+        assert expr.shape == self.value_shape
+
+        scalar_i = self.base_element.get_indices()
+        scalar_vi = self.base_element.get_value_indices()
+        tensor_i = tuple(gem.Index(extent=d) for d in self._shape)
+        tensor_vi = tuple(gem.Index(extent=d) for d in self._shape)
+
+        if self._transpose:
+            index_ordering = tensor_i + scalar_i + tensor_vi + scalar_vi
         else:
-            num_factors = 1
-        Q_indexed = Q[Q_shape_indices[:num_factors] + base_value_indices + expr_contraction_indices + basis_tensor_indices]
-        expr_indexed = expr[expr_contraction_indices + base_value_indices]
-        dual_evaluation_indexed_sum = gem.optimise.make_product((Q_indexed, expr_indexed), x.indices + base_value_indices + expr_contraction_indices)
-        basis_indices = Q_shape_indices[:num_factors] + basis_tensor_indices
-        return dual_evaluation_indexed_sum, basis_indices
+            index_ordering = scalar_i + tensor_i + tensor_vi + scalar_vi
+
+        Qi = Q[index_ordering]
+        expri = expr[tensor_i + scalar_vi]
+        evaluation = gem.IndexSum(Qi * expri, x.indices + scalar_vi + tensor_i)
+        evaluation = gem.optimise.contraction(evaluation, tensor_vi)
+        return evaluation, scalar_i + tensor_vi
 
     @property
     def mapping(self):
```

Reply:
That doesn't work (the old code also not) if the `TensorElement` was made with `symmetry=True` (so that `element.value_shape() != element.reference_value_shape()`

Reply:
yes I couldn't work out how to make this work in general. I've added the below as documentation for the tensor finite element case:
```python
    @property
    def dual_basis(self):
        '''Return a dual evaluation gem weight tensor Q and point set x to dual
        evaluate a function fn at.
...
        If the dual basis is of a tensor finite element with some shape 
        (S1, S2, ..., Sn) then the tensor element Q is constructed from the 
        base element's Q (here called q) by taking the outer product with 
        appropriately sized identity matrices:
        Q = q ⊗ 𝟙ₛ₁ ⊗ 𝟙ₛ₂ ⊗ ... ⊗ 𝟙ₛₙ
        '''
```

Reply:
> That doesn't work (the old code also not) if the `TensorElement` was made with `symmetry=True` (so that `element.value_shape() != element.reference_value_shape()`

When does that happen? No where in the firedrake test suite covers that

Reply:
If I make a `TensorElement("P", triangle, 1, symmetry=True)`.

Reply:
But I think the problem is actually not in this code, but rather that in TSFC we need to apply a non-affine mapping to the expression we compile in the case of a symmetries=True TensorElement.

Reply:
It looks like FInAT elements don't have a property called reference_value_shape - any ideas on how I'd raise a `NotImplementedError` or `ValueError` in FInAT for this case?

Reply:
Can you just do it in Firedrake?
```patch
diff --git a/firedrake/interpolation.py b/firedrake/interpolation.py
index c5db09f6..3ff0974d 100644
--- a/firedrake/interpolation.py
+++ b/firedrake/interpolation.py
@@ -236,6 +236,8 @@ def _interpolator(V, tensor, expr, subset, arguments, access):
     try:
         if not isinstance(expr, firedrake.Expression):
             to_element = create_element(V.ufl_element())
+            if V.ufl_element().mapping() == "symmetries":
+                raise NotImplementedError("Interpolation into tensor spaces with symmetry")
         else:
             # compile_python_kernel code pathway expects base element
             to_element = create_base_element(V.ufl_element())
```

Reply:
I think I broadly know what needs to be done, but a bunch of refactoring is required.

Reply:
> Maybe we have to have the dual_evaluation separate for this case. Perhaps this diff, which I think handles the `transpose` argument correctly?
> 
> ```diff
> diff --git a/finat/tensorfiniteelement.py b/finat/tensorfiniteelement.py
> index 839a009..9535a6e 100644
> --- a/finat/tensorfiniteelement.py
> +++ b/finat/tensorfiniteelement.py
> @@ -5,8 +5,6 @@ import numpy
>  import gem
>  
>  from finat.finiteelementbase import FiniteElementBase
> -from finat.tensor_product import total_num_factors
> -from finat.cube import FlattenedDimensions
>  
>  
>  class TensorFiniteElement(FiniteElementBase):
> @@ -124,59 +122,51 @@ class TensorFiniteElement(FiniteElementBase):
>  
>      @property
>      def dual_basis(self):
> -
>          base = self.base_element
>          Q, points = base.dual_basis
>  
>          # Suppose the tensor element has shape (2, 4)
>          # These identity matrices may have difference sizes depending the shapes
>          # tQ = Q ⊗ 𝟙₂ ⊗ 𝟙₄
> -        Q_shape_indices = gem.indices(len(Q.shape))
> -        i_s = tuple(gem.Index(extent=d) for d in self._shape)
> -        j_s = tuple(gem.Index(extent=d) for d in self._shape)
> -        # we need one delta for each shape axis
> -        deltas = reduce(gem.Product, (gem.Delta(i, j) for i, j in zip(i_s, j_s)))
> -        # TODO Need to check how this plays with the transpose argument to TensorFiniteElement.
> -        tQ = gem.ComponentTensor(Q[Q_shape_indices]*deltas, Q_shape_indices + i_s + j_s)
> +        scalar_i = self.base_element.get_indices()
> +        scalar_vi = self.base_element.get_value_indices()
> +        tensor_i = tuple(gem.Index(extent=d) for d in self._shape)
> +        tensor_vi = tuple(gem.Index(extent=d) for d in self._shape)
> +        # Couple new basis function and value indices
> +        deltas = reduce(gem.Product, (gem.Delta(j, k)
> +                                      for j, k in zip(tensor_i, tensor_vi)))
> +        if self._transpose:
> +            index_ordering = tensor_i + scalar_i + tensor_vi + scalar_vi
> +        else:
> +            index_ordering = scalar_i + tensor_i + tensor_vi + scalar_vi
>  
> -        return tQ, points
> +        Qi = Q[scalar_i + scalar_vi]
> +        Q = gem.ComponentTensor(Qi*deltas, index_ordering)
> +        return Q, points
>  
>      def dual_evaluation(self, fn):
> -
>          Q, x = self.dual_basis
>          expr = fn(x)
> -
> -        # TODO: Add shortcut (if relevant) for Q being identity tensor
> -
> +        gem.optimise.contraction(expr)
>          # NOTE: any shape indices in the expression are because the expression
>          # is tensor valued.
> -        assert set(expr.shape) == set(self.value_shape)
> -
> -        base_value_indices = self.base_element.get_value_indices()
> -
> -        # Reconstruct the indices from the deltas in dual_basis above and
> -        # contract Q with expr
> -        Q_shape_indices = tuple(gem.Index(extent=ex) for ex in Q.shape)
> -        expr_contraction_indices = tuple(gem.Index(extent=d) for d in self._shape)
> -        basis_tensor_indices = tuple(gem.Index(extent=d) for d in self._shape)
> -        # NOTE: here the first num_factors rows of Q are node sets (1 for
> -        # each factor)
> -        # TODO: work out if there are any other cases where the basis
> -        # indices in the shape of the dual basis tensor Q are more than
> -        # just the first shape index (e.g. with EnrichedElement)
> -        if hasattr(self.base_element, 'factors'):
> -            num_factors = total_num_factors(self.base_element.factors)
> -        elif isinstance(self.base_element, FlattenedDimensions):
> -            # Factor might be a FlattenedDimensions which introduces
> -            # another factor without having a factors property
> -            num_factors = 2
> +        assert expr.shape == self.value_shape
> +
> +        scalar_i = self.base_element.get_indices()
> +        scalar_vi = self.base_element.get_value_indices()
> +        tensor_i = tuple(gem.Index(extent=d) for d in self._shape)
> +        tensor_vi = tuple(gem.Index(extent=d) for d in self._shape)
> +
> +        if self._transpose:
> +            index_ordering = tensor_i + scalar_i + tensor_vi + scalar_vi
>          else:
> -            num_factors = 1
> -        Q_indexed = Q[Q_shape_indices[:num_factors] + base_value_indices + expr_contraction_indices + basis_tensor_indices]
> -        expr_indexed = expr[expr_contraction_indices + base_value_indices]
> -        dual_evaluation_indexed_sum = gem.optimise.make_product((Q_indexed, expr_indexed), x.indices + base_value_indices + expr_contraction_indices)
> -        basis_indices = Q_shape_indices[:num_factors] + basis_tensor_indices
> -        return dual_evaluation_indexed_sum, basis_indices
> +            index_ordering = scalar_i + tensor_i + tensor_vi + scalar_vi
> +
> +        Qi = Q[index_ordering]
> +        expri = expr[tensor_i + scalar_vi]
> +        evaluation = gem.IndexSum(Qi * expri, x.indices + scalar_vi + tensor_i)
> +        evaluation = gem.optimise.contraction(evaluation, tensor_vi)
> +        return evaluation, scalar_i + tensor_vi
>  
>      @property
>      def mapping(self):
> ```

This is great, much neater than what I wrote. What's the reason for `gem.optimise.contraction(expr)`? Also I presume you meant to write `evaluation = gem.optimise.contraction(evaluation)`?

Reply:
I'll prepare a PR into this branch...

Reply:
FYI I've already added the above - just wanted to annotate it with some comments. I've pushed what I did